p2p/src/authenticated/discovery/actors/peer/actor.rs

```diff
@@ -200,14 +203,41 @@ impl<E: Spawner + Clock + Rng + CryptoRng + Metrics, C: PublicKey> Actor<E, C> {
 
                     // Update metrics
                     let metric = match &msg {
+                        types::Payload::Greeting(_) => &metrics::Message::new_greeting(&peer),
                         types::Payload::BitVec(_) => &metrics::Message::new_bit_vec(&peer),
                         types::Payload::Peers(_) => &metrics::Message::new_peers(&peer),
                         types::Payload::Data(data) => &metrics::Message::new_data(&peer, data.channel),
                     };
                     self.received_messages.get_or_create(metric).inc();
 
+                    // Ensure we start with a greeting message and then never receive another
+                    if let types::Payload::Greeting(info) = msg {
+                        if greeting_received {
+                            debug!(?peer, "received duplicate greeting");
+                            return Err(Error::DuplicateGreeting);
+                        }
+                        greeting_received = true;
+
+                        // Verify the greeting is from the expected peer
+                        if info.public_key != peer {
```

Comment:
This check is new (we weren't doing this on main over the first peers message received (but probably should've)